flink-test-utils-parent/flink-connector-testing-framework/src/main/java/org/apache/flink/connectors/test/common/testsuites/TestSuiteBase.java

```diff
@@ -0,0 +1,430 @@
+     *
+     * <p>This test will create 1 split in the external system, write test record set A into the
+     * split, restart task manager to trigger job failover, write test record set B into the split,
+     * and terminate the Flink job finally.
+     *
+     * <p>The number and order of records consumed by Flink should be identical to A before the
+     * failover and B after the failover in order to pass the test.
+     *
+     * <p>An unbounded source is required for this test, since TaskManager failover will be
+     * triggered in the middle of the test.
+     */
+    @Case
+    @Tag("failover")
+    @DisplayName("Test TaskManager failure")
+    public void testTaskManagerFailure(TestEnvironment testEnv, ExternalContext<T> externalContext)
+            throws Exception {
+
+        checkEnvironmentIsControllable(testEnv);
+
+        final Collection<T> testRecordsBeforeFailure = externalContext.generateTestData();
+        final SourceSplitDataWriter<T> sourceSplitDataWriter = externalContext.createSourceSplit();
+        sourceSplitDataWriter.writeRecords(testRecordsBeforeFailure);
+
+        final StreamExecutionEnvironment env = testEnv.createExecutionEnvironment();
+
+        env.enableCheckpointing(50);
+        final DataStreamSource<T> dataStreamSource =
+                env.fromSource(
+                                externalContext.createSource(Boundedness.CONTINUOUS_UNBOUNDED),
+                                WatermarkStrategy.noWatermarks(),
+                                "Tested Source")
+                        .setParallelism(1);
+
+        // Since DataStream API doesn't expose job client for executeAndCollect(), we have
+        // to reuse these part of code to get both job client and result iterator :-(
+        // ------------------------------------ START ---------------------------------------------
+        TypeSerializer<T> serializer = dataStreamSource.getType().createSerializer(env.getConfig());
+        String accumulatorName = "dataStreamCollect_" + UUID.randomUUID();
+        CollectSinkOperatorFactory<T> factory =
+                new CollectSinkOperatorFactory<>(serializer, accumulatorName);
+        CollectSinkOperator<T> operator = (CollectSinkOperator<T>) factory.getOperator();
+        CollectResultIterator<T> iterator =
+                new CollectResultIterator<>(
+                        operator.getOperatorIdFuture(),
+                        serializer,
+                        accumulatorName,
+                        env.getCheckpointConfig());
+        CollectStreamSink<T> sink = new CollectStreamSink<>(dataStreamSource, factory);
+        sink.name("Data stream collect sink");
+        env.addOperator(sink.getTransformation());
+        final JobClient jobClient = env.executeAsync("TaskManager Failover Test");
+        iterator.setJobClient(jobClient);
+        // -------------------------------------- END ---------------------------------------------
```

Comment:
I don't get why you can't use `DataStream#executeAndCollect`. Could you elaborate?

Reply:
```DataStream#executeAndCollect``` doesn't return ```JobClient```, which is required in ```triggerTaskManagerFailure```. There's a method ```DataStream#executeAndCollectWithClient``` that can fulfill the requirement, but unfortunately this method is not public. Maybe we can make it public in another ticket (or FLIP since ```DataStream``` is public API)

Reply:
I think we should make it public in this PR in a separate commit. @sjwiesman , do you know why this is currently not exposed?
I'm assuming a FLIP is not needed as the other executeAndCollect methods have also been added without FLIP.

Reply:
Maybe mark it as `@Internal`? We kept it package-private because we didn't believe it was a method any normal user should use.

Reply:
I agree that the method may be too complex for end-users. I think we should leave things as is until we have a better idea. Adding `@Internal` on 1 method in a public API feels weird to me.